<launch>
  <arg name="first_tb3"  default="tb3_0"/>
  <arg name="second_tb3" default="tb3_1"/>
  <arg name="dim_state"  default="6"/>
  <arg name="marker_length"  default="0.22884"/>
  <arg name="aruco_cal_file"  default="calibrationSave_2.p"/>
  <arg name="first_tb3_id"    default=" 2001"/>
  <arg name="second_tb3_id"    default=" 2002"/>

  <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_world.py" name="dse_gazebo_object_visualization_node" output="screen">
    <param name="n_params" value="2" />
    <rosparam param="objects"> ["aruco_marker_0", "aruco_marker_1"]</rosparam>
  </node>

  <node pkg="dse_simulation" type="centralized_estimator_10hz_2.py" name="dse_centralized_estimator_node" output="screen">
    <param name="n_params" value="2" />
    <rosparam param="objects" subst_value="True"> [$(arg first_tb3), $(arg second_tb3)]</rosparam>
    <param name="dim_state" type="int" value="$(arg dim_state)" />
  </node>

  <group ns = "$(arg first_tb3)">
    <node pkg="dse_simulation" type="aruco_pose_estimation" name="aruco_pose_estimation_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="marker_length" type="double" value="$(arg marker_length)" />
      <param name="calibration_file" value="$(arg aruco_cal_file)" />
    </node>
    <node pkg="dse_simulation" type="information_filter" name="information_filter_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="rqt_plot_output.py" name="dse_plotting_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_base.py" name="dse_gazebo_visualization_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="world_tf_broadcaster.py" name="world_tf_broadcaster" output="screen">
      <param name="gazebo_name" value="$(arg first_tb3)" />
      <param name="tf_name" value="$(arg first_tb3)/odom" />
    </node>
    <node pkg="dse_simulation" type="circle_controller.py" name="circle_controller_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="fwd_vel" type="double" value="0.1" />
    </node>
  </group>

  <group ns = "$(arg second_tb3)">
    <node pkg="dse_simulation" type="aruco_pose_estimation" name="aruco_pose_estimation_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="marker_length" type="double" value="$(arg marker_length)" />
      <param name="calibration_file" value="$(arg aruco_cal_file)" />
    </node>
    <node pkg="dse_simulation" type="information_filter" name="information_filter_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="rqt_plot_output.py" name="dse_plotting_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_base.py" name="dse_gazebo_visualization_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="world_tf_broadcaster.py" name="world_tf_broadcaster" output="screen">
      <param name="gazebo_name" value="$(arg second_tb3)" />
      <param name="tf_name" value="$(arg second_tb3)/odom" />
    </node>
    <node pkg="dse_simulation" type="circle_controller.py" name="circle_controller_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="fwd_vel" type="double" value="0.1" />
    </node>
  </group>
</launch>
